Writerside/topics/implementation/public-transit-service.md

```diff
@@ -1,9 +1,166 @@
+   
+              for (int i = key.length() - 1; i >= 0; i--) {
+                  suffixTrie.insert(key.substring(i), entry);
+              }
+   
+          }
+   
+      }
+      ```
+
+Using these indices, queries for stops are highly efficient. The trade-off, however, is that these data structures
+consume more memory.
 
 ## Connection Routing Service
 
-RoutePartitioner
+The `ConnectionRoutingService` is responsible for finding connections between transit stops. In our `GtfsRaptorService`
+implementation, this service interacts with the [extended RAPTOR routing engine](extended-raptor.md), which processes
+routing requests based on the GTFS schedule data.
+
+### Route Partitioning
+
+The GTFS model and RAPTOR algorithm have differing perspectives on routes. In GTFS, a "route" refers to a user-friendly
+view of trips that may share a common line number or route name, even if the stop sequences differ slightly. For
+instance, a regional train might have trips that skip certain stops on later trips while still being part of the same
+route from a user's point of view.
+
+However, RAPTOR requires that all trips associated with a route must have the exact same stop sequence. This discrepancy
+is resolved by the `GtfsToRaptorConverter` by applying the `GtfsRoutePartitioner`, which splits GTFS routes into "
+sub-routes". These sub-routes group trips that share identical stop sequences, ensuring compatibility with RAPTOR’s
+routing model. During post-processing, when the routing results are returned to the user, the sub-routes are converted
+back into the original GTFS routes to maintain user-friendly responses (TODO: Where is this happening? in the RAPTOR or
+in the service?).
+
+### Transfer Generation
+
+In GTFS, the transfer file is optional, meaning not all transfers between stops may be defined. To handle this, we
+introduced an optional transfer generation step in the `GtfsToRaptorConverter`. This step uses
+the `WalkTransferGenerator` to generate walking transfers between stops that are within a reasonable walking distance
+but are missing from the GTFS transfer data. Even when GTFS transfer data is provided, the generator can add missing,
+but viable, transfers and thus improve the connection results.
+
+The walking distances between stops are calculated using a `WalkCalculator`. This implementation uses a **beeline
+distance factor**, typically set to 1.3 based on literature [13, p. 43], to account for real-world walking conditions
+such as obstacles, detours, or building access routes. The calculator multiplies the straight-line distance by this
+factor to approximate a more realistic walking time. The `WalkCalculator` interface allows for easy replacement with
+more advanced footpath routing solutions, such as A* or A+ Landmark algorithms, if needed.
+
+### Handling Coordinates
+
+In routing requests, users can specify coordinates (latitude/longitude) or stops as the origin or destination. When
```

Comment:
This is actually also a point we can add to the extended raptor, that we modified / extended the algorithm to allow setting multiple source stops (with different source times) and target multiple target stops (with different handicaps)

Reply:
Good point, I added a TODO.